src/python/pants/backend/python/typecheck/mypy/rules.py

```diff
@@ -227,6 +237,58 @@ async def mypy_typecheck_partition(
         requirements_venv_pex_request, file_list_digest_request
     )
 
+    py_version = config_file.python_version_to_autoset(
+        partition.interpreter_constraints, python_setup.interpreter_versions_universe
+    )
+    named_cache_dir = f".cache/mypy_cache/{sha256(build_root.path.encode()).hexdigest()}"
+    run_cache_dir = ".tmp_cache/mypy_cache"
+    argv = await _generate_argv(
+        mypy,
+        pex=mypy_pex,
+        venv_python=requirements_venv_pex.python.argv0,
+        cache_dir=run_cache_dir,
+        file_list_path=file_list_path,
+        python_version=config_file.python_version_to_autoset(
+            partition.interpreter_constraints, python_setup.interpreter_versions_universe
+        ),
+    )
+
+    script_runner_digest = await Get(
+        Digest,
+        CreateDigest(
+            [
+                FileContent(
+                    "__run_wrapper.sh",
+                    dedent(
+                        f"""\
+                            # We want to leverage the MyPy cache for fast incremental runs of MyPy.
+                            # Pants exposes "append_only_caches" we can leverage, but with the caveat
+                            # that it requires either only appending files, or multiprocess-safe access.
+                            #
+                            # MyPy guaruntees neither, but there's workarounds!
+                            #
+                            # By default, MyPy uses 2 cache files persource file, which introduces a
+                            # whole slew of race conditions. To workaround this we enable MyPy's SQLite cache,
+                            # which enables us to run MyPy on a temporary copy of the cache, then mv
+                            # the modified cache back to the real location. This is multiprocess-safe
+                            # as mv on the same filesystem is an atomic "rename", and any processes
+                            # copying the "old" file will still have valid file descriptors for the
+                            # "old" file.
+
+                            {mkdir.path} -p {run_cache_dir}/{py_version} 2>&1 > /dev/null
+                            {cp.path} {named_cache_dir}/{py_version}/cache.db {run_cache_dir}/{py_version}/cache.db 2>&1 > /dev/null || true
```

Comment:
This assumed `py_version` is not `None`, which isn't always the case. We have two options:

1. Turn off cache support if the user provided a python version (and amend the error message to mention that they are also might be losing caching)
2. Yank support for user-configured `python_version` value. From the error: "...This feature may no longer work."

I think we should do 1, but wanted to ask explicitly in PR what others thought.

(this is a blocker for this PR)

Reply:
It wasn't too bad to just turn it off and improve the message